src/corehost/cli/hostpolicy/hostpolicy_context.cpp

```diff
@@ -146,27 +146,42 @@ int hostpolicy_context_t::initialize(hostpolicy_init_t &hostpolicy_init, const a
     for (int i = 0; i < hostpolicy_init.cfg_keys.size(); ++i)
     {
         // Provide opt-in compatible behavior by using the switch to set APP_PATHS
-        if (pal::strcasecmp(hostpolicy_init.cfg_keys[i].c_str(), _X("Microsoft.NETCore.DotNetHostPolicy.SetAppPaths")) == 0)
+        const pal::char_t *key = hostpolicy_init.cfg_keys[i].c_str();
+        if (pal::strcasecmp(key, _X("Microsoft.NETCore.DotNetHostPolicy.SetAppPaths")) == 0)
         {
             set_app_paths = (pal::strcasecmp(hostpolicy_init.cfg_values[i].data(), _X("true")) == 0);
         }
 
-        coreclr_properties.add(hostpolicy_init.cfg_keys[i].c_str(), hostpolicy_init.cfg_values[i].c_str());
+        if (!coreclr_properties.add(key, hostpolicy_init.cfg_values[i].c_str()))
+        {
+            trace::error(_X("Duplicate runtime property found: %s"), key);
+            return StatusCode::LibHostDuplicateProperty;
+        }
     }
 
     // App paths and App NI paths.
     // Note: Keep this check outside of the loop above since the _last_ key wins
     // and that could indicate the app paths shouldn't be set.
     if (set_app_paths)
     {
-        coreclr_properties.add(common_property::AppPaths, app_base.c_str());
-        coreclr_properties.add(common_property::AppNIPaths, app_base.c_str());
+        if (!coreclr_properties.add(common_property::AppPaths, app_base.c_str())
+            || !coreclr_properties.add(common_property::AppNIPaths, app_base.c_str()))
+        {
+            trace::error(_X("Duplicate runtime property for app paths"));
```

Comment:
Please include the exact name of the affected properties - it's easier to debug then.
I would also add a single line of explanation, something like: "It's invalid to specify values for properties populated by the hosting layer in the application's .runtimeconfig.json."